File: vca.lua
--------------------------------------------------------------------------------
-- VOLTAGE-CONTROLLED AMPLIFIER
--------------------------------------------------------------------------------

-- knobs
vca_env = 0
vca_env_ar = true
vca_gain = 0

-- out
vca_out = 0

function update_vca()
    local env = vca_env_ar and ar_out or adsr_out
    vca_out = (vca_env * env + vca_gain) * hpf_out
end

File: vca.cpp
////////////////////////////////////////////////////////////////////////////////
// VOLTAGE-CONTROLLED AMPLIFIER
////////////////////////////////////////////////////////////////////////////////

// in
extern float
ar_out,
adsr_out,
hpf_out;

// knobs
float
vca_env = 0,
vca_gain = 0;
bool
vca_env_ar = true;

// out
float
vca_out = 0;

void update_vca() {
    auto env = vca_env_ar ? ar_out : adsr_out;
    vca_out = (vca_env * env + vca_gain) * hpf_out;
}
